File: lib/services/sync_service.dart
import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/history_record.dart';
import 'user_service.dart';
import 'storage_service.dart';

/// 同步服务 - 云同步、多设备数据互通
class SyncService {
  static final SyncService _instance = SyncService._internal();
  factory SyncService() => _instance;
  SyncService._internal();

  final Dio _dio = Dio(BaseOptions(
    baseUrl: 'https://your-backend-url.com', // TODO: 替换为实际后端地址
    connectTimeout: const Duration(seconds: 30),
    receiveTimeout: const Duration(seconds: 30),
    headers: {'Content-Type': 'application/json'},
  ));

  static const String _baseUrlPrefKey = 'cleartalk_api_base_url';

  final UserService _userService = UserService();
  StorageService? _storageService;
  SharedPreferences? _prefs;

  // 同步状态
  bool _isSyncing = false;
  DateTime? _lastSyncTime;

  // 状态监听器
  final List<Function(bool)> _syncListeners = [];

  /// 初始化
  Future<void> init() async {
    _prefs = await SharedPreferences.getInstance();

    final savedBaseUrl = _prefs?.getString(_baseUrlPrefKey);
    if (savedBaseUrl != null && savedBaseUrl.trim().isNotEmpty) {
      _dio.options.baseUrl = savedBaseUrl.trim();
    }

    _lastSyncTime = _prefs?.getInt('last_sync_time') != null
        ? DateTime.fromMillisecondsSinceEpoch(_prefs!.getInt('last_sync_time')!)
        : null;

    // 设置 Dio 拦截器
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) {
        final token = _userService.token;
        if (token != null) {
          options.headers['Authorization'] = 'Bearer $token';
        }
        handler.next(options);
      },
    ));
  }

  /// 设置存储服务
  void setStorageService(StorageService storage) {
    _storageService = storage;
  }

  /// 是否正在同步
  bool get isSyncing => _isSyncing;

  /// 上次同步时间
  DateTime? get lastSyncTime => _lastSyncTime;

  /// 添加同步状态监听
  void addListener(Function(bool) listener) {
    _syncListeners.add(listener);
  }

  /// 移除同步状态监听
  void removeListener(Function(bool) listener) {
    _syncListeners.remove(listener);
  }

  /// 通知监听器
  void _notifyListeners(bool syncing) {
    for (final listener in _syncListeners) {
      listener(syncing);
    }
  }

  /// 上传到云端
  Future<Map<String, dynamic>> upload() async {
    if (!_userService.isLoggedIn) {
      throw Exception('Login required');
    }

    if (_storageService == null) {
      throw Exception('Storage service not initialized');
    }

    try {
      // 收集本地数据（与本地备份逻辑保持一致）
      final data = await _storageService!.exportForSync();

      final response = await _dio.post('/api/sync/upload', data: {'data': data});

      // 保存同步时间
      _lastSyncTime = DateTime.now();
      await _prefs?.setInt('last_sync_time', _lastSyncTime!.millisecondsSinceEpoch);

      return {
        'success': true,
        'timestamp': _lastSyncTime,
        'serverTime': response.data['timestamp'],
      };
    } on DioException catch (e) {
      final message = e.response?.data?['error'] ?? e.message ?? 'Network error';
      throw Exception(message);
    }
  }

  /// 从云端下载
  Future<Map<String, dynamic>> download() async {
    if (!_userService.isLoggedIn) {
      throw Exception('Login required');
    }

    try {
      final response = await _dio.get('/api/sync/download');
      final data = response.data;

      return {
        'success': true,
        'history': (data?['history'] as List<dynamic>?)?.map((json) {
          try {
            return HistoryRecord.fromJson(json);
          } catch (_) {
            return null;
          }
        }).whereType<HistoryRecord>().toList() ?? [],
        'customScenes': data?['customScenes'] ?? [],
        'settings': data?['settings'] ?? {},
        'phrases': List<String>.from(data?['phrases'] ?? []),
        'quickPhrases': List<String>.from(data?['quickPhrases'] ?? []),
        'lastModified': data?['lastSync'] != null
            ? DateTime.tryParse(data['lastSync'].toString())
            : null,
      };
    } on DioException catch (e) {
      final message = e.response?.data?['error'] ?? e.message ?? 'Network error';
      throw Exception(message);
    }
  }

  /// 同步到云端（立即上传）
  Future<Map<String, dynamic>> syncToCloud() async {
    _isSyncing = true;
    _notifyListeners(true);

    try {
      final result = await upload();
      _isSyncing = false;
      _notifyListeners(false);
      return {'success': true, 'timestamp': result['timestamp']};
    } catch (error) {
      _isSyncing = false;
      _notifyListeners(false);
      return {'success': false, 'error': error.toString()};
    }
  }

  /// 从云端同步（下载并合并）
  Future<Map<String, dynamic>> syncFromCloud({bool merge = true}) async {
    if (!_userService.isLoggedIn) {
      return {'success': false, 'error': 'Not logged in'};
    }

    _isSyncing = true;
    _notifyListeners(true);

    try {
      final cloud = await download();

      if (_storageService == null) {
        throw Exception('Storage service not initialized');
      }

      final payload = <String, dynamic>{
        'history': (cloud['history'] as List<HistoryRecord>).map((r) => r.toJson()).toList(),
        'customScenes': cloud['customScenes'],
        'settings': cloud['settings'],
        'phrases': cloud['phrases'],
        'quickPhrases': cloud['quickPhrases'],
      };

      await _storageService!.importFromSync(payload, merge: merge);

      // 更新同步时间
      _lastSyncTime = DateTime.now();
      await _prefs?.setInt('last_sync_time', _lastSyncTime!.millisecondsSinceEpoch);

      _isSyncing = false;
      _notifyListeners(false);

      return {
        'success': true,
        'timestamp': _lastSyncTime,
        'recordsCount': (cloud['history'] as List).length,
      };
    } catch (error) {
      _isSyncing = false;
      _notifyListeners(false);
      return {'success': false, 'error': error.toString()};
    }
  }

  /// 自动同步（登录后调用）
  Future<void> autoSync() async {
    if (!_userService.isLoggedIn) return;

    // 检查网络状态（实际实现需要 connectivity_plus）
    try {
      // 优先从云端下载合并
      await syncFromCloud(merge: true);
      // 然后上传本地新数据
      await syncToCloud();
    } catch (e) {
      // 自动同步失败不抛出错误
      print('Auto sync failed: $e');
    }
  }

  /// 获取同步状态文本
  String getSyncStatusText() {
    if (_isSyncing) return 'syncing';
    if (_lastSyncTime == null) return 'never';
    
    final diff = DateTime.now().difference(_lastSyncTime!);
    if (diff.inMinutes < 1) return 'just_now';
    if (diff.inHours < 1) return '${diff.inMinutes} minutes ago';
    if (diff.inDays < 1) return '${diff.inHours} hours ago';
    return '${diff.inDays} days ago';
  }

  /// 设置后端地址
  void setBaseUrl(String url) {
    _dio.options.baseUrl = url;
    _prefs?.setString(_baseUrlPrefKey, url);
  }

  String get baseUrl => _dio.options.baseUrl;
}

File: src/services/sync-service.js
/**
 * 同步服务 - 云同步、多设备数据互通
 */

import { api } from './api-client.js';
import { Storage } from '../core/storage.js';
import { state } from '../core/state.js';
import { logger } from '../core/logger.js';
import { UserService } from './user-service.js';
import { events, deepMerge } from '../core/utils.js';
import { t } from '../core/i18n.js';

export const SyncService = {
  // 上传到云端
  async upload() {
    if (!UserService.isLoggedIn()) {
      throw new Error(t('auth.error.login.required'));
    }
    
    const data = {
      history: Storage.getHistory(),
      customScenes: Storage.getCustomScenes(),
      settings: Storage.get('cleartalk_settings', {}),
      phrases: Storage.get('cleartalk_phrases', []),
      dialogueContexts: Storage.get('cleartalk_dialogue_contexts', {})
    };
    
    const result = await api.post('/api/sync/upload', { data });
    
    // 保存同步时间
    Storage.setSetting('lastSyncTime', Date.now());
    events.emit('sync:upload', result);
    
    return result;
  },
  
  // 从云端下载
  async download() {
    if (!UserService.isLoggedIn()) {
      throw new Error(t('auth.error.login.required'));
    }
    
    const result = await api.get('/api/sync/download');
    
    return {
      history: result.data?.history || [],
      customScenes: result.data?.customScenes || [],
      settings: result.data?.settings || {},
      phrases: result.data?.phrases || [],
      dialogueContexts: result.data?.dialogueContexts || {},
      lastModified: result.lastModified
    };
  },
  
  // 同步到云端（立即上传）
  async syncToCloud() {
    try {
      const result = await this.upload();
      return { success: true, timestamp: result.timestamp };
    } catch (error) {
      logger.error('Sync to cloud failed:', error);
      return { success: false, error: error.message };
    }
  },
  
  // 合并云端和本地数据
  async mergeFromCloud() {
    if (!UserService.isLoggedIn()) {
      throw new Error(t('auth.error.login.required'));
    }
    
    state.isSyncing = true;
    
    try {
      const cloud = await this.download();
      const local = {
        history: Storage.getHistory(),
        customScenes: Storage.getCustomScenes()
      };
      
      // 合并历史记录（去重）
      const mergedHistory = this._mergeHistory(local.history, cloud.history);
      
      // 合并自定义场景（云端优先）
      const mergedScenes = this._mergeScenes(local.customScenes, cloud.customScenes);
      
      // 保存到本地
      Storage.set('cleartalk_history', mergedHistory);
      Storage.set('cleartalk_custom_scenes', mergedScenes);
      
      // 合并设置
      if (Object.keys(cloud.settings).length > 0) {
        const mergedSettings = deepMerge(
          Storage.get('cleartalk_settings', {}),
          cloud.settings
        );
        Storage.set('cleartalk_settings', mergedSettings);
      }
      
      // 保存同步时间
      Storage.setSetting('lastSyncTime', Date.now());
      
      events.emit('sync:merge', {
        historyCount: mergedHistory.length,
        scenesCount: mergedScenes.length
      });
      
      return {
        historyCount: mergedHistory.length,
        scenesCount: mergedScenes.length,
        addedHistory: mergedHistory.length - local.history.length,
        addedScenes: mergedScenes.length - local.customScenes.length
      };
    } finally {
      state.isSyncing = false;
    }
  },
  
  // 下载并覆盖本地
  async downloadAndReplace() {
    if (!UserService.isLoggedIn()) {
      throw new Error(t('auth.error.login.required'));
    }
    
    state.isSyncing = true;
    
    try {
      const cloud = await this.download();
      
      // 直接覆盖
      if (cloud.history.length > 0) {
        Storage.set('cleartalk_history', cloud.history);
      }
      if (cloud.customScenes.length > 0) {
        Storage.set('cleartalk_custom_scenes', cloud.customScenes);
      }
      if (Object.keys(cloud.settings).length > 0) {
        Storage.set('cleartalk_settings', cloud.settings);
      }
      if (cloud.phrases.length > 0) {
        Storage.set('cleartalk_phrases', cloud.phrases);
      }
      
      Storage.setSetting('lastSyncTime', Date.now());
      
      events.emit('sync:replace', cloud);
      
      return {
        historyCount: cloud.history.length,
        scenesCount: cloud.customScenes.length
      };
    } finally {
      state.isSyncing = false;
    }
  },
  
  // 清除云端数据
  async clearCloud() {
    if (!UserService.isLoggedIn()) {
      throw new Error(t('auth.error.login.required'));
    }
    
    await api.delete('/api/sync/clear');
    events.emit('sync:clear');
    return { success: true };
  },
  
  // 获取同步状态
  getStatus() {
    const lastSync = Storage.getSetting('lastSyncTime', 0);
    const autoSync = Storage.getSetting('autoSync', false);
    const direction = Storage.getSetting('syncDirection', 'merge');
    
    return {
      lastSync,
      lastSyncFormatted: lastSync ? new Date(lastSync).toLocaleString('zh-CN') : '从未',
      autoSync,
      direction,
      isSyncing: state.isSyncing,
      canSync: UserService.isLoggedIn()
    };
  },
  
  // 配置自动同步
  setAutoSync(enabled) {
    Storage.setSetting('autoSync', enabled);
    return { success: true };
  },
  
  // 配置同步方向
  setDirection(direction) {
    const validDirections = ['merge', 'upload', 'download'];
    if (!validDirections.includes(direction)) {
      throw new Error(t('settings.sync.invalid.direction'));
    }
    Storage.setSetting('syncDirection', direction);
    return { success: true };
  },
  
  // 执行配置的同步
  async syncWithDirection() {
    const direction = this.getStatus().direction;
    
    switch (direction) {
      case 'merge':
        return await this.mergeFromCloud();
      case 'upload':
        await this.upload();
        return { uploaded: true };
      case 'download':
        return await this.downloadAndReplace();
      default:
        return await this.mergeFromCloud();
    }
  },
  
  // 内部：合并历史记录
  _mergeHistory(local, cloud) {
    const map = new Map();
    
    // 先添加本地
    local.forEach(item => {
      map.set(item.id, item);
    });
    
    // 合并云端（以时间较新的为准）
    cloud.forEach(item => {
      const existing = map.get(item.id);
      if (!existing) {
        map.set(item.id, item);
      } else {
        const localTime = new Date(existing.createdAt || 0).getTime();
        const cloudTime = new Date(item.createdAt || 0).getTime();
        if (cloudTime > localTime) {
          map.set(item.id, item);
        }
      }
    });
    
    // 转回数组，按时间倒序
    return Array.from(map.values()).sort((a, b) => {
      return new Date(b.createdAt || 0) - new Date(a.createdAt || 0);
    });
  },
  
  // 内部：合并场景
  _mergeScenes(local, cloud) {
    const map = new Map();
    
    // 先添加云端
    cloud.forEach(scene => {
      map.set(scene.id, scene);
    });
    
    // 合并本地（本地新增的场景保留）
    local.forEach(scene => {
      if (!map.has(scene.id)) {
        map.set(scene.id, scene);
      }
    });
    
    return Array.from(map.values());
  }
};

export default SyncService;
